### ISSUE TITLE
qemu: Bump to version 5.2.0

### ISSUE THREAD

**Issue by wainersm** [kata-containers/kata-containers#1238]
**Is your feature request related to a problem? Please describe.**

virtiofsd from currently supported QEMU (5.0.0) crashes on Red Hat CoreOS. As a result, Kata Containers configured to use the virtio-fs sharing cannot be tested on OpenShift.

**Describe the solution you'd like**

The fix (https://github.com/qemu/qemu/commit/ebf101955ce8f8d72fba103b5151115a4335de2c) was released on QEMU 5.2.0, therefore I am suggesting here to bump the QEMU version which is supported/tested on CI to 5.2.0.

**Describe alternatives you've considered**

Unless QEMU 5.2.0 is not possible/desirable at this point, alternatively we can:

  1. Back-port the commit ebf101955ce8f8d72fba103 to QEMU 5.0.0. However the code base is too different and it is likely we will need to back-port various other commits 

  2. Bump to 5.1.0 instead of 5.2.0 and back-port the commit. On 5.1.0 the commit can be simple cherry-picked without conflicts. 

  3. Not test Kata with virtiofsd on OpenShift. Not a solution at all.

**Additional context**

I did an initial assessment on the work needed to update the version. It is needed to:

 1. QEMU employs a new build system (Meson) which depends on Ninja build. So the ninja-build package should be installed on the Ubuntu container.

 2. Do not build with `-O3` flag. Meson warns that it should be used "the default optimization flag" instead. And, in fact, with -O3 a object on the migration module fails on compilation.

 3.The virtiofsd make target (`make -j$(nproc) virtiofsd`) does not exist. In order to build the daemon binary it will need to pass `--enable-virtiofsd` configure flag. 

Things what should be investigated:

 1. Whether `--enable-virtiofsd` is really the way to enable virtiofsd build

 2. Should we apply the other 5.x.0 patches? (https://github.com/kata-containers/kata-containers/tools/packaging/qemu/patches/5.0.x)

 3. Should the metrics CI values adjusted?
 
 4. Should be bump kernel too?

**Work in progress**

The final patch can be evolved from:

```
commit daf8cecc338e98de3b1b1c59128a89187202a90a (HEAD -> qemu_5_2)
Author: Wainer dos Santos Moschetta <wainersm@redhat.com>
Date:   Wed Jan 6 06:53:27 2021 -0500

    [wip] Bump to QEMU 5.2.0
    
    Signed-off-by: Wainer dos Santos Moschetta <wainersm@redhat.com>

diff --git a/tools/packaging/scripts/configure-hypervisor.sh b/tools/packaging/scripts/configure-hypervisor.sh
index 195d25b3..d1714a23 100755
--- a/tools/packaging/scripts/configure-hypervisor.sh
+++ b/tools/packaging/scripts/configure-hypervisor.sh
@@ -445,7 +445,9 @@ generate_qemu_options() {
        _qemu_cflags=""
 
        # compile with high level of optimisation
-       _qemu_cflags+=" -O3"
+       if [ "${qemu_version_major}" -le 4 ] || [[ "${qemu_version_major}" -eq 5 && "${qemu_version_minor}" -lt 2 ]]; then
+               _qemu_cflags+=" -O3"
+       fi
 
        # Improve code quality by assuming identical semantics for interposed
        # synmbols.
diff --git a/tools/packaging/static-build/qemu/Dockerfile b/tools/packaging/static-build/qemu/Dockerfile
index 502b9018..3d678753 100644
--- a/tools/packaging/static-build/qemu/Dockerfile
+++ b/tools/packaging/static-build/qemu/Dockerfile
@@ -40,6 +40,7 @@ RUN apt-get --no-install-recommends install -y \
            libselinux1-dev \
            libtool \
            make \
+           ninja-build \
            pkg-config \
            libseccomp-dev \
            libseccomp2 \
@@ -65,6 +66,5 @@ RUN PREFIX="${PREFIX}" /root/configure-hypervisor.sh -s kata-qemu | xargs ./conf
        --with-pkgversion=kata-static
 
 RUN make -j$(nproc)
-RUN make -j$(nproc) virtiofsd
 RUN make install DESTDIR="${QEMU_DESTDIR}"
 RUN /root/static-build/scripts/qemu-build-post.sh
diff --git a/versions.yaml b/versions.yaml
index 6b33e626..3cb2c08b 100644
--- a/versions.yaml
+++ b/versions.yaml
@@ -88,8 +88,8 @@ assets:
     qemu:
       description: "VMM that uses KVM"
       url: "https://github.com/qemu/qemu"
-      version: "5.0.0"
-      tag: "v5.0.0"
+      version: "5.2.0"
+      tag: "v5.2.0"
       # Do not include any non-full release versions
       # Break the line *without CR or space being appended*, to appease
       # yamllint, and note the deliberate ' ' at the end of the expression.
```

**Comment by wainersm** [comment]
Hi @amshinde , could you please let me know if this update cannot be carried out for some reason which is unknown to me? If it's okay, could you also tell me what else need to be changed to update the qemu version? Thanks!

**Comment by jodh-intel** [comment]
Hi @wainersm - You seem to have raised two almost identical issues? Can you close #1237?

> virtiofsd from currently supported QEMU (5.0.0) crashes on Red Hat CoreOS

Why is it crashing? Does it crash on RHEL 8 / CentOS 8? Is there a bug you can refer us to?

Regarding updating QEMU, we obviously wish to stay current with new releases so backporting upstream patches to older upstream releases is something to avoid. However, we still like to get a feel for how new releases compare with what we're currently using. To do this, we run the [metrics report](https://github.com/kata-containers/tests/tree/master/metrics/report) and provide the results. tal at https://github.com/kata-containers/runtime/pull/2355 for an excellent example from @devimc. Depending on those results, yes, we may need to adjust the metrics CI - @GabyCT is probably the person to advise on this.

In terms of the packaging work, the configure flags will need to be adjusted in either https://github.com/kata-containers/kata-containers/blob/2.0-dev/tools/packaging/scripts/configure-hypervisor.sh or https://github.com/kata-containers/kata-containers/tree/2.0-dev/tools/packaging/static-build/qemu - @jcvenegas should be able to commeont on this next week. It sounds rather worrying about the `-O3` compile flag (regression). meson is rather an unknown - it might be easy to setup (?) but will obviously take effort to ensure everything works for the CI and metrics builds.

**Comment by wainersm** [comment]
> Hi @wainersm - You seem to have raised two almost identical issues? Can you close #1237?
> 

Sorry, closed that one.

> > virtiofsd from currently supported QEMU (5.0.0) crashes on Red Hat CoreOS
> 
> Why is it crashing? Does it crash on RHEL 8 / CentOS 8? Is there a bug you can refer us to?
> 

That's the bug: https://bugzilla.redhat.com/show_bug.cgi?id=1880932

IIUC virtiofsd used to try to mount `/proc/self/fd` on a tmpfs. On Red Hat CoreOS (RHCOS) some portions of the file system are read-only (related with immutable principles employed), so that the mount operation was failing with "operation not permitted". The problem cannot be reproduced easily on rhel or centos because those don't use the immutable file system mechanisms.

**Comment by jing-wang4** [comment]
This [fix](https://github.com/dagrh/qemu/commit/dcaac9f124833e62e720d0733aa53453f438b7d7) is needed for ppc64le. Shall we open another issue for backport or bump? 

cc: @bpradipt @Amulyam24 @clnperez

**Comment by clnperez** [comment]
@jing-wang4 I recall someone mentioning that it is fixed in 5.2, so correct me if I'm mistaken there. 

To elaborate a bit on that request, we'd either want to bump up to requiring 5.2 (for a different reason than the one in the issue, but maybe we can just piggyback), or ask for a backport of that into qemu 5.0 (the currently supported/recommended qemu version for kata).

**Comment by wainersm** [comment]
@jing-wang4 @clnperez I am in favour of bump to 5.2.

BTW, I just added this as a topic for tomorrow's (Jan 26) Kata Containers Architecture Committee Meeting -> https://etherpad.opendev.org/p/Kata_Containers_2021_Architecture_Committee_Mtgs

**Comment by wainersm** [comment]
@jcvenegas @jodh-intel Do you know why we compile QEMU with `-O3`?

I'd like to re-visit that decision because when I build QEMU 5.2 with Kata's configuration:

1) Meson says it should not be used -O3 (no further information is provided)
2) It fails to compile `migration/qemu-file.c` due a "maybe uninitialized" error. I believe that `-O3` implicitly enables `-Werror=maybe-uninitialized` so resulting on that error. I sent a patch to fix the code  (see https://lists.gnu.org/archive/html/qemu-devel/2021-01/msg07262.html) but, even if accepted, we will need to apply it manually on our build tree and I'd like to avoid carrying patches on our side.

**Comment by wainersm** [comment]
Last week I figured out the `virtiofsd` make target no longer exists on 5.2 (and newer). It gets built with `--enable-tools --enable-virtiofsd` but we don't want to build the entire tools set (e.g, qemu-img ...) since 1) might fail the static build; 2) increase the attach surface; 2) they aren't used by kata at all.

So I am working on a way to build only virtiofsd.

**Comment by wainersm** [comment]
> Last week I figured out the `virtiofsd` make target no longer exists on 5.2 (and newer). It gets built with `--enable-tools --enable-virtiofsd` but we don't want to build the entire tools set (e.g, qemu-img ...) since 1) might fail the static build; 2) increase the attach surface; 2) they aren't used by kata at all.
> 
> So I am working on a way to build only virtiofsd.

And here is the patch I sent to QEMU-devel: https://lists.gnu.org/archive/html/qemu-devel/2021-02/msg00291.html

**Comment by wainersm** [comment]
Now I am figuring out what patches from 5.0 and 5.1 should be also applied.

From `packaging/qemu/patches/5.0.x`:

 * 0001-9p-removing-coroutines-of-9p-to-increase-the-I-O-per.patch
  This patch is mentioned on https://www.redhat.com/archives/virtio-fs/2020-November/msg00094.html (commit e49393a349925567cb8 - virtiofsd: Use --thread-pool-size=0 to mean no thread pool) but I could not find a fix for 9p. @jcvenegas  Should we apply or drop it?
 * 0002-memory-backend-file-nvdimm-support-read-only-files-a.patch
   This was implemented as a nvdimm feature (https://mail.gnu.org/archive/html/qemu-devel/2021-01/msg00258.html) but not merged yet. @egernst is it going to require any change on Go side? (@fidencio FYI)
 * 0003-virtiofsd-Used-glib-shared-thread-pool.patch
   Don't need this as the patch is applied on QEMU 5.2 already (commit 04d325e86f7)

**Comment by wainersm** [comment]
From packaging/qemu/patches/5.0.x:

 * 0001-9p-removing-coroutines-of-9p-to-increase-the-I-O-per.patch
    This patch is mentioned on https://www.redhat.com/archives/virtio-fs/2020-November/msg00094.html (commit e49393a349925567cb8 - virtiofsd: Use --thread-pool-size=0 to mean no thread pool) but I could not find a fix for 9p. @jcvenegas Should we apply or drop it?
  * 0002-memory-backend-file-nvdimm-support-read-only-files-a.patch
    This was implemented as a nvdimm feature (https://mail.gnu.org/archive/html/qemu-devel/2021-01/msg00258.html) and merged today on commit 8360ebeb4f4a7079. @egernst is it going to require any change on Go side? (@fidencio FYI)
  * 0003-virtiofsd-Used-glib-shared-thread-pool.patch
    Don't need this as the patch is applied on QEMU 5.2 already (commit 04d325e86f7)

**Comment by wainersm** [comment]
@devimc Julio, what's the purpose of the QEMU configuration files under https://github.com/kata-containers/kata-containers/tree/main/tools/packaging/qemu/default-configs ? Should they be updated too?

**Comment by devimc** [comment]
@wainersm this configuration file is used to reduce the number of default devices available in the final binary (optimization), so yes, you should update it according to the new list of default devices

**Comment by fidencio** [comment]
@wainersm, please, add the patches from https://lists.gnu.org/archive/html/qemu-devel/2021-02/msg01787.html as part of the bump.

**Comment by jcvenegas** [comment]
@wainersm kata 2.0 does not use 9pfs so should OK drop it

**Comment by fidencio** [comment]
@wainersm, I'm commeting here as I got to this issue when going through all the issues with the "needs-review" label.  I've removed the label as this is a valid issue, and I've added the "area/packaging" label.

If you feel something is not correct, please, add the most correct labels. :-)

**Comment by wainersm** [comment]
> @wainersm, please, add the patches from https://lists.gnu.org/archive/html/qemu-devel/2021-02/msg01787.html as part of the bump.

Done.

**Comment by wainersm** [comment]
> @wainersm kata 2.0 does not use 9pfs so should OK drop it

@jcvenegas ah, ok. so I won't apply qemu/patches/5.0.x/0001-9p-removing-coroutines-of-9p-to-increase-the-I-O-per.patch on 5.2.0.

**Comment by wainersm** [comment]
@devimc Julio, still about https://github.com/kata-containers/kata-containers/tree/main/tools/packaging/qemu/default-configs I don't see those files being used by the scripts to build the static QEMU. Could you explain me where/how they are used?

**Comment by wainersm** [comment]
I checked the patches from ```qemu/patches/5.1.x```; some are already merged in QEMU 5.2.0, others will be applied from ```qemu/patches/5.2.x```. So if we decide to update the aarch64 configuration (in ```versions.yaml```) to use QEMU 5.2.0 too, then I think it will work just fine.

**Comment by wainersm** [comment]
> This [fix: virtiofsd: Seccomp: Add 'send' for syslog](https://github.com/dagrh/qemu/commit/dcaac9f124833e62e720d0733aa53453f438b7d7) is needed for ppc64le. Shall we open another issue for backport or bump?
> 

@bpradipt @Amulyam24 @clnperez @jing-wang4 guys, that patch is already merged on 5.2.0. Is there any other patch needed for kata containers on ppc64le?

**Comment by Amulyam24** [comment]
> > This [fix: virtiofsd: Seccomp: Add 'send' for syslog](https://github.com/dagrh/qemu/commit/dcaac9f124833e62e720d0733aa53453f438b7d7) is needed for ppc64le. Shall we open another issue for backport or bump?
> 
> @bpradipt @Amulyam24 @clnperez @jing-wang4 guys, that patch is already merged on 5.2.0. Is there any other patch needed for kata containers on ppc64le?

This should be good, thanks @wainersm!

**Comment by Jakob-Naucke** [comment]
@jongwu since 5.2.0 is removing support for 9p: Is virtio-fs working on ARM? It wasn't working according to #990 and I had similar symptoms on s390x

**Comment by jongwu** [comment]
thanks @Jakob-Naucke -, virtio-fs can work on arm64 for kernel 5.6+, so the kernel for arm64 should upgrade and I need to investigate for qemu-5.2. Since 9p will be deprecated we should hurry up @justin-he .

**Comment by wainersm** [comment]
> @jongwu since 5.2.0 is removing support for 9p: Is virtio-fs working on ARM? It wasn't working according to #990 and I had similar symptoms on s390x

@Jakob-Naucke @jongwu ,

I'm not sure I understood. QEMU 5.2.0 has still support for 9p afaik, same for Kata Containers 2.x. The different on later is that virtio-fs is now the default backend.

There is a patch kept on kata containers's repo which is meant for QEMU 5.0.0 's 9p. I was thinking in drop it, but given that you guys are concerned about ARM, I am better to apply it on 5.0 as well.

**Comment by jongwu** [comment]
@wainersm Good. Thanks for your clarification.

**Comment by Jakob-Naucke** [comment]
@wainersm you're right, QEMU 5.2.0 isn't dropping 9p. I was hastily judging all of QEMU by my s390x experience :) It seems the `virtio-9p-ccw` device (required for s390x) is not included in QEMU 5.2.0 anymore (at least by default, maybe I've missed a config flag. I'll check.) Yes, please do keep the 9p patches for the time being.
